### PR TITLE
Basic support for Rocky Linux 8

### PR DESCRIPTION
**Pull request by digdilem**
## What does this PR change?

Add support for Rocky Linux as a supported distro.

Packages building for tests on top of Uyuni Master: https://build.opensuse.org/project/show/home:juliogonzalezgil:branches:systemsmanagement:Uyuni:Master

Java code: https://github.com/uyuni-project/uyuni/pull/3959

# Tests 

## WebUI, regular minion

- [x] Bootstrap
- [x] Hardware refresh
- [x] Enabling a formula (locale) and adjusting data
- [x] Applying the highstate with the formula
- [x] Detect pending updates
- [x] Install a package
- [x] Remove a package
- [ ] Errata detection: Not available -> https://forums.rockylinux.org/t/will-there-be-errata/1806
- [x] Run a remote command

## WebUI, SSH minion

- [x] Bootstrap
- [x] Hardware refresh
- [x] Enabling a formula (locale) and adjusting data
- [x] Applying the highstate with the formula
- [x] Detect pending updates
- [ ] Install a package: Fails, known issue (https://github.com/SUSE/spacewalk/issues/14400). The GPG key for Rocky must be trusted before the onboarding.
- [x] Remove a package
- [ ] Errata detection: Not available -> https://forums.rockylinux.org/t/will-there-be-errata/1806
- [x] Run a remote command

## Bootstrap script, regular minion

- [x] Bootstrap
- [x] Hardware refresh
- [x] Enabling a formula (locale) and adjusting data
- [x] Applying the highstate with the formula
- [x] Detect pending updates
- [x] Install a package
- [x] Remove a package
- [ ] Errata detection: Not available -> https://forums.rockylinux.org/t/will-there-be-errata/1806
- [x] Run a remote command

## GUI diff

No difference.

## Documentation
- Pending (https://github.com/SUSE/spacewalk/issues/14220)

### CODE REVIEW

File: utils/spacewalk-common-channels.ini [hunk 1]
```diff
@@ -2949,3 +2949,59 @@ gpgkey_url = https://download.opensuse.org/repositories/systemsmanagement:/Uyuni
 gpgkey_id = %(_uyuni_gpgkey_id)s
 gpgkey_fingerprint = %(_uyuni_gpgkey_fingerprint)s
 repo_url = https://download.opensuse.org/repositories/systemsmanagement:/Uyuni:/Master:/CentOS8-Uyuni-Client-Tools/CentOS_8/
+
+
+[rockylinux8]
+archs    = x86_64
+name     = RockyLinux 8 (%(arch)s)
+gpgkey_url = !!!!!!!!- Unable to locate
+gpgkey_id = !!!!!!- Unable to locate
```
Comment (paususe) [inline utils/spacewalk-common-channels.ini:2958]: Keys can be found in the Rocky git repository for the time being but I have asked the Rocky team to make them available from download.rockylinux.org, which is the proper place. They will.

https://git.rockylinux.org/staging/rpms/rocky-release/-/tree/r8/SOURCES

Comment (paususe) [reply utils/spacewalk-common-channels.ini:2958]: And added they are:
https://dl.rockylinux.org/pub/rocky/

Comment (digdilem) [reply utils/spacewalk-common-channels.ini]: Thanks @paususe - amended commit with updated info.

url as given (official, not testing)
id from: gpg --import rocky.gpg
Fingerprint from: gpg --show-keys rocky.gpg

File: spacewalk/certs-tools/spacewalk-certs-tools.changes [hunk 2]
```diff
@@ -1,3 +1,9 @@
+-------------------------------------------------------------------
+Sun May 2 14:08:00 GMT 2021 digdilem@gmail.com
+- Prepare the bootstrap script generator for Rockylinux 8
+
+
+-------------------------------------------------------------------
```
Comment (mcalmer) [inline spacewalk/certs-tools/spacewalk-certs-tools.changes:6]: For changelogs please add just the line with the change on top of the file.
The "Header" will be added by our tooling when we submit the change to the build service.
```suggestion
- Prepare the bootstrap script generator for Rockylinux 8
```

File: susemanager-utils/susemanager-sls/susemanager-sls.changes [hunk 3]
```diff
@@ -1,3 +1,10 @@
+-------------------------------------------------------------------
+Sun May 2nd 18:33:00 BST - digdilem@gmail.com
+
+- Add Rocky Linux support
+
+-------------------------------------------------------------------
+
```
Comment (mcalmer) [inline susemanager-utils/susemanager-sls/susemanager-sls.changes:7]:
```suggestion
- Add Rocky Linux support
```
Same as above

File: susemanager-utils/susemanager-sls/salt/bootstrap/init.sls [hunk 4]
```diff
@@ -36,6 +36,9 @@ mgr_server_localhost_alias_absent:
 {%- elif salt['file.file_exists']('/etc/alinux-release') %}
 {% set bootstrap_repo_url = 'https://' ~ salt['pillar.get']('mgr_server') ~ '/pub/repositories/alibaba/' ~ grains['osmajorrelease'] ~ '/bootstrap/' %}
 
+{%- elif salt['file.file_exists']('/etc/rocky-release') %}
+{% set bootstrap_repo_url = 'https://' ~ salt['pillar.get']('mgr_server') ~ '/pub/repositories/rocky/' ~ grains['osmajorrelease'] ~ '/bootstrap/' %}
```
Comment (mcalmer) [inline susemanager-utils/susemanager-sls/salt/bootstrap/init.sls:40]: I think the path to the bootstrap repo is wrong.

In spacewalk/certs-tools/rhn_bootstrap_strings.py you set `BASE` to "rockylinux" which match the path in "susemanager/src/mgr_bootstrap_data.py " where you define the bootstrap repo at: `/srv/www/htdocs/pub/repositories/rockylinux/8/bootstrap/` . 
Just "rocky" is not enough here.

```suggestion
{% set bootstrap_repo_url = 'https://' ~ salt['pillar.get']('mgr_server') ~ '/pub/repositories/rockylinux/' ~ grains['osmajorrelease'] ~ '/bootstrap/' %}
```

File: susemanager/susemanager.changes [hunk 5]
```diff
@@ -1,3 +1,9 @@
+-------------------------------------------------------------------
+Sun May 2nd 18:35:00 BST - digdilem@gmail.com
+
+- Add bootstrap repository definitions for RockyLinux8
+-------------------------------------------------------------------
+
```
Comment (mcalmer) [inline susemanager/susemanager.changes:6]:
```suggestion
- Add bootstrap repository definitions for RockyLinux8
```

File: utils/spacewalk-utils.changes [hunk 6]
```diff
@@ -1,3 +1,8 @@
+Sun May 2nd 18:47:00 BST - digdilem@gmail.com
+
+- Add Rocky Linux repo info
+
+-------------------------------------------------------------------
```
Comment (mcalmer) [inline utils/spacewalk-utils.changes:5]:
```suggestion
- Add Rocky Linux repo info
```

**Review by mcalmer — CHANGES_REQUESTED** [review]

**Comment by lumarel** [comment]
I was already in need for this patch, so I manually applied it.
There I noticed some quirks which aren't fully correct in this patch up to now, I hope I am allowed to point them out.

File: susemanager/src/mgr_bootstrap_data.py [hunk 7]
```diff
@@ -1311,6 +1311,10 @@
     'almalinux-8-x86_64-uyuni' : {
         'BASECHANNEL' : 'almalinux8-x86_64', 'PKGLIST' : RES8 + RES8_X86,
         'DEST' : '/srv/www/htdocs/pub/repositories/almalinux/8/bootstrap/'
+    },    
+    'rockylinux-8-x86_64-uyuni' : {
+        'BASECHANNEL' : 'rockylinux-x86_64', 'PKGLIST' : RES8 + RES8_X86,
```
Comment (lumarel) [inline susemanager/src/mgr_bootstrap_data.py:1316]: The basechannel should actually be rockylinux8-x86_64 or it won't be able to create the bootstrap repo.

File: utils/spacewalk-common-channels.ini [hunk 8]
```diff
@@ -2949,3 +2949,59 @@ gpgkey_url = https://download.opensuse.org/repositories/systemsmanagement:/Uyuni
...
+gpgkey_url = https://dl.rockylinux.org/pub/rocky/RPM-GPG-KEY-rockyofficial
+gpgkey_id = 6D745A60
+gpgkey_fingerprint = 7051 C470 A929 F454 CEBE 37B7 15AF 5DAC 6D74 5A60
+repo_url = https://dl.rockylinux.org/pub/rocky/8/BaseOS/%(arch)s/os/
+dist_map_release = 8
+
+[rockylinux8-appstream]
+label    = %(base_channel)s-appstream
+archs    = x86_64
+name     = rockyLinux 8 AppStream (%(arch)s)
```
Comment (lumarel) [inline utils/spacewalk-common-channels.ini:2966]: In this and the following sections the first letter of the name is written in lowercase, which is a little bit ugly.
(Btw the official way the repos are written as "Rocky Linux $releasever - <repo>")

File: utils/spacewalk-common-channels.ini [hunk 9]
```diff
@@ -2949,3 +2949,59 @@ gpgkey_url = https://download.opensuse.org/repositories/systemsmanagement:/Uyuni
...
 repo_url = https://download.opensuse.org/repositories/systemsmanagement:/Uyuni:/Master:/CentOS8-Uyuni-Client-Tools/CentOS_8/
+
+
+[rockylinux8]
+archs    = x86_64
+name     = RockyLinux 8 (%(arch)s)
+gpgkey_url = https://dl.rockylinux.org/pub/rocky/RPM-GPG-KEY-rockyofficial
+gpgkey_id = 6D745A60
+gpgkey_fingerprint = 7051 C470 A929 F454 CEBE 37B7 15AF 5DAC 6D74 5A60
+repo_url = https://dl.rockylinux.org/pub/rocky/8/BaseOS/%(arch)s/os/
```
Comment (lumarel) [inline utils/spacewalk-common-channels.ini:2960]: I think, here and in the following sections, should really be used the mirrorlist, to get better performance out of the box.
Yes the Tier 0 Mirror `dl.rockylinux.org` is a CDN, but this has a limited capacity in some locations (and isn't even available in others)
So in this case it should be `https://mirrors.rockylinux.org/mirrorlist?repo=BaseOS-8&arch=%(arch)s`

File: susemanager-utils/susemanager-sls/salt/certs/Rocky8.sls [hunk 10]
```diff
@@ -0,0 +1 @@
+RedHat8.sls
```
Comment (lumarel) [inline susemanager-utils/susemanager-sls/salt/certs/Rocky8.sls:1]: This does not work, it has to have the whole content of the RedHat8.sls file in it, which means:
```
/etc/pki/ca-trust/source/anchors/RHN-ORG-TRUSTED-SSL-CERT:
  file.managed:
    - source:
      - salt://certs/RHN-ORG-TRUSTED-SSL-CERT

update-ca-certificates:
  cmd.run:
    - name: /usr/bin/update-ca-trust extract
    - runas: root
    - onchanges:
      - file: /etc/pki/ca-trust/source/anchors/RHN-ORG-TRUSTED-SSL-CERT
```

Comment (digdilem) [reply susemanager-utils/susemanager-sls/salt/certs/Rocky8.sls:1]: Apologies - how do I generate this in github?

File: utils/spacewalk-common-channels.ini (hunk 8 above)
Comment (digdilem) [reply utils/spacewalk-common-channels.ini]: Good suggestion, implemented

File: susemanager-utils/susemanager-sls/salt/certs/Rocky8.sls (hunk 10 above)
Comment (lumarel) [reply susemanager-utils/susemanager-sls/salt/certs/Rocky8.sls:1]: Hmm... good question, I would have done it via a bash with `ln -s` but I either can't find a function directly in Github for that

**Comment by juliogonzalez** [comment]
@digdilem can you give me write access to your fork, so I can contribute changes there? In general the PR looks OK, but I want to test everything, fix the issue with the link to the SSL certificate and other things, maybe add the Java support as well.

**Comment by digdilem** [comment]
> @digdilem can you give me write access to your fork, so I can contribute changes there? In general the PR looks OK, but I want to test everything, fix the issue with the link to the SSL certificate and other things, maybe add the Java support as well.

@juliogonzalez  - Done, invite sent. Please advise if I need to do anything else.

File: susemanager/susemanager.changes [hunk 11]
```diff
@@ -25,7 +26,7 @@ Fri Apr 16 13:27:26 CEST 2021 - jgonzalez@suse.com
 
 -------------------------------------------------------------------
 Fri Mar 05 15:43:37 CET 2021 - jgonzalez@suse.com
-
+
```
Comment (juliogonzalez) [inline susemanager/susemanager.changes:29]: This is not needed.

File: utils/spacewalk-common-channels.ini (hunk 9 above)
Comment (juliogonzalez) [reply utils/spacewalk-common-channels.ini:2960]: I will fix it. I requested @digdilem permission to write on his fork, to do the fixes for the PR, and maybe some enhacements.

**Comment by juliogonzalez** [comment]
> @juliogonzalez - Done, invite sent. Please advise if I need to do anything else.

Works fine. I tested with a commit to the branch :-)

Thanks!

File: susemanager-utils/susemanager-sls/salt/certs/Rocky8.sls (hunk 10 above)
Comment (juliogonzalez) [reply susemanager-utils/susemanager-sls/salt/certs/Rocky8.sls]: Fixed.

File: susemanager/src/mgr_bootstrap_data.py (hunk 7 above)
Comment (juliogonzalez) [reply susemanager/src/mgr_bootstrap_data.py]: Seems to be fixed.

File: utils/spacewalk-common-channels.ini (hunk 9 above)
Comment (juliogonzalez) [reply utils/spacewalk-common-channels.ini]: Fixed (pending testing).

File: susemanager/susemanager.changes (hunk 11 above)
Comment (juliogonzalez) [reply susemanager/susemanager.changes]: Fixed.

File: utils/spacewalk-common-channels.ini (hunk 1 above)
Comment (juliogonzalez) [reply utils/spacewalk-common-channels.ini]: Seems this was fixed (pending tests).

**Comment by juliogonzalez** [comment]
I rebased the changes from master, fixed some things with the `spacewalk-common-channels` definitions, and fixed the changelogs. For now this should be enough for basic support. Time to create an environment and see how this works (to be done today/tomorrow).

@digdilem JFYI: the rebase rewrote the git history, so if you want to pull the changes, you need to perform a hard resent with `git reset --hard origin/master` (assuming `origin` is the remote for your fork)

**Review by paususe — APPROVED** [review]
LGTM

**Comment by juliogonzalez** [comment]
@digdilem can you have a look at Gitter? I left you a private message for you there :-)

**Pull request merged**